### DISCUSSION TITLE
How to avoid multiple database calls when populating user info in the session callback

### DISCUSSION THREAD

**Discussion by ChrisB1123** [nextauthjs/next-auth#9321] [Help]
**Goal** 
Populate my session object with a user's ID from my database while using JWT session strategy.

**Background**
I didn't mind using database sessions, which I guess would solve this issue, but I get this error when doing so:
>Error: PrismaClient is unable to run in Vercel Edge Functions. As an alternative, try Accelerate: https://pris.ly/d/accelerate.

I don't believe it's possible to turn off Edge Functions here? So I switched to a JWT strategy.

**Problem**
Since the JWT strategy only populates the session with what you get from the Provider, I need to connect to my database to fetch the userId whenever the session is populated.

I have this set up and working fine, however if I put a console.log in my API route, it's getting called 3 times in production, and more in development. I'd like to just make this call once (or just make the call to the database once, if Prisma can memoise the call in some way)

This is my auth config:
```
auth.ts

export const config = {
  ...
  callbacks: {
    ...
    async session({ session, token, user }) {
      if (session.user && session.user.email) {
        const userResponse = await fetch('http://localhost:3000/api/auth/user?' + new URLSearchParams({
            email: session.user.email
        }))
        const foundUser = await userResponse.json()

        if (foundUser) {
          session.user.id = foundUser.id;
        }
      }
      return session
    }
  },
  providers: [...],
} 
```

and here is my API route:
```
app/api/auth/user/route.ts

export async function GET(request: Request) {
  const { searchParams } = new URL(request.url)
  const email = searchParams.get('email')

  console.log('hello')

  if (!email) {
    return Response.json({ error: 'email is required' })
  }

  const foundUser = await prisma.user.findUnique({ where: { email: email } });
  return Response.json(foundUser)

}
```

**Stack**
- "next": "14.0.3",
- "next-auth": "^5.0.0-beta.4",
- "@prisma/client": "^5.6.0",
- "prisma": "^5.6.0",
- Heroku/Postgres

**Comment by vniehues** [comment]
Hi @ChrisB1123 did you ever find a solution for this?

**Comment by eddedd88** [comment]
Since you are using Prisma, you should use the prisma adapter to handle this for you [(docs)](https://authjs.dev/getting-started/adapters/prisma#old-edge-workaround), it uses **next-auth v5 (beta)**; and make sure to extend the session with the user id from the jwt [(docs)](https://authjs.dev/guides/extending-the-session#with-jwt). So the auth config would look something like this:
```typescript
import NextAuth from "next-auth"
import { PrismaAdapter } from "@auth/prisma-adapter"
import { PrismaClient } from "@prisma/client"
 
globalThis.prisma ??= new PrismaClient()
 
export const { handlers, auth } = NextAuth({
  adapter: PrismaAdapter(globalThis.prisma),
  session: { strategy: "jwt" },

  // add the user id to the session
  callbacks: {
    jwt({ token, user }) {
      if (user) { // User is available during sign-in
        token.id = user.id
      }
      return token
    },
    session({ session, token }) {
      session.user.id = token.id
      return session
    },
  },
  ...
})
```

If you want to stick to a `database` session strategy, **Prisma** recently released support for Edge functions, [see here](https://www.prisma.io/docs/orm/prisma-client/deployment/edge/overview).
